### PR TITLE
[P1-M01] Separate round snapshot function

### PR DESCRIPTION
**Pull request by chrismaree**
This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run.  

This PR introduces a second option in snapshotting by directly calling a `snapshotCurrentRound` function which will capture a snapshot in a seperate tx to the first revealer. This gives the first revealer the _option_ to separate their reveal into two transactions and prevent any front running.

Note: The original means of snapshotting on first reveal is still possible. This is left up to the first revealer as the hit on UX on first reveal is very large and so by having both options the choice is given to the first revealer.

### CODE REVIEW

**Comment by mrice32** [comment]
> This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.

Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?

**Comment by chrismaree** [comment]
> > This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.
> 
> Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?

I originally did it like actually where either could work: you can call this new function to snapshot or you can be the first revealer which will snapshot. The problem is that in all cases the first reveal can be frontrun, unless it is a completely separate transaction. If we dont built the off chain infrastructure what we can do is check if a snapshot has been made on the front end and then send two txs from the wallet of the first revealer.

**Comment by mrice32** [comment]
> > > This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.
> > 
> > 
> > Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?
> 
> I originally did it like actually where either could work: you can call this new function to snapshot or you can be the first revealer which will snapshot. The problem is that in all cases the first reveal can be frontrun, unless it is a completely separate transaction. If we dont built the off chain infrastructure what we can do is check if a snapshot has been made on the front end and then send two txs from the wallet of the first revealer.

I think it's a question of whether we want you to be able to optionally run this transaction separately or make it mandatory.

**Comment by chrismaree** [comment]
> > > > This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.
> > > 
> > > 
> > > Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?
> > 
> > 
> > I originally did it like actually where either could work: you can call this new function to snapshot or you can be the first revealer which will snapshot. The problem is that in all cases the first reveal can be frontrun, unless it is a completely separate transaction. If we dont built the off chain infrastructure what we can do is check if a snapshot has been made on the front end and then send two txs from the wallet of the first revealer.
> 
> I think it's a question of whether we want you to be able to optionally run this transaction separately or make it mandatory.

For sure, we could definitely enable the option and let the user choose. We can give the user revealing a warning in the UI if they are the first to reveal. 

My logic in choosing to do it this was to be consistent so that in all cases we are never susceptible to front running of any kind. You have to sign two transactions often when smart contract systems. think an approve then trade on uniswap for example, so it's not that uncommon. 

if we are fine with the front running then we can enable both logic to run such that it's an either or.

**Comment by mrice32** [comment]
> > > > > This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.
> > > > 
> > > > 
> > > > Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?
> > > 
> > > 
> > > I originally did it like actually where either could work: you can call this new function to snapshot or you can be the first revealer which will snapshot. The problem is that in all cases the first reveal can be frontrun, unless it is a completely separate transaction. If we dont built the off chain infrastructure what we can do is check if a snapshot has been made on the front end and then send two txs from the wallet of the first revealer.
> > 
> > 
> > I think it's a question of whether we want you to be able to optionally run this transaction separately or make it mandatory.
> 
> For sure, we could definitely enable the option and let the user choose. We can give the user revealing a warning in the UI if they are the first to reveal.
> 
> My logic in choosing to do it this was to be consistent so that in all cases we are never susceptible to front running of any kind. You have to sign two transactions often when smart contract systems. think an approve then trade on uniswap for example, so it's not that uncommon.
> 
> if we are fine with the front running then we can enable both logic to run such that it's an either or.

@ptare @nicholaspai thoughts?

**Comment by nicholaspai** [comment]
> > > > > > This PR changes how snapshotting works in the DVM. Previously the first revealer would execute the snapshot. This opens the first revealer up to being front run. This PR moves this snapshotting process to a seperate function call to prevent front running. This does place a nock on the UX if calling this function is not automated. Therefore we should consider placing some off-chain infrastructure that will call this function automatically for us.
> > > > > 
> > > > > 
> > > > > Should we make it optional to call this function before the reveal? Put another way, should we call this function inside the reveal if it hasn't been called externally yet so we don't hurt the UX?
> > > > 
> > > > 
> > > > I originally did it like actually where either could work: you can call this new function to snapshot or you can be the first revealer which will snapshot. The problem is that in all cases the first reveal can be frontrun, unless it is a completely separate transaction. If we dont built the off chain infrastructure what we can do is check if a snapshot has been made on the front end and then send two txs from the wallet of the first revealer.
> > > 
> > > 
> > > I think it's a question of whether we want you to be able to optionally run this transaction separately or make it mandatory.
> > 
> > 
> > For sure, we could definitely enable the option and let the user choose. We can give the user revealing a warning in the UI if they are the first to reveal.
> > My logic in choosing to do it this was to be consistent so that in all cases we are never susceptible to front running of any kind. You have to sign two transactions often when smart contract systems. think an approve then trade on uniswap for example, so it's not that uncommon.
> > if we are fine with the front running then we can enable both logic to run such that it's an either or.
> 
> @ptare @nicholaspai thoughts?

I think most importantly we should prevent front-running the snapshot. It seems like this will have to add a second sign-transaction prompt to the UX. Said another way, if there weren't two separate transactions, then there would be possibility for front-running.

**Comment by ptare** [comment]
I'm pretty indifferent, don't really think the front running is a super serious issue.

File: core/contracts/oracle/implementation/Voting.sol [hunk 1]
```diff
@@ -374,7 +389,8 @@ contract Voting is Testable, Ownable, OracleInterface, VotingInterface, Encrypte
         )) == voteSubmission.commit, "Revealed data != commit hash");
         delete voteSubmission.commit;
 
-        // Lock in round variables including snapshotId and inflation rate
+        // Lock in round variables including snapshotId and inflation rate. Not that this will only execute a snapshot
+        // if the `snapshotCurrentRound` function was not already called for this round.
         _freezeRoundVariables(roundId);
```
Comment (chrismaree) [inline core/contracts/oracle/implementation/Voting.sol:394]: This is now exactly as it was before. You can now snapshot as the first revealer _or_ by calling `snapshotCurrentRound`.

**Comment by chrismaree** [comment]
@mrice32, @ptare  and @nicholaspai I've updated this PR to include both options of snapshotting.

File: core/contracts/oracle/implementation/Voting.sol [hunk 2]
```diff
@@ -339,10 +339,25 @@ contract Voting is Testable, Ownable, OracleInterface, VotingInterface, Encrypte
         emit VoteCommitted(msg.sender, currentRoundId, identifier, time);
     }
 
+    /**
+     * @notice Snapshot the current round's token balances and lock in the inflation rate and GAT.
+     * @dev This function can be called multiple times but each round will only every have one snapshot at the
+     * time of calling `_freezeRoundVariables`.
```
Comment (mrice32) [inline core/contracts/oracle/implementation/Voting.sol:345]:
```suggestion
     * @dev This function can be called multiple times, but only the first call per round into this function or `revealVote`
     * will create the round snapshot. Any later calls will be a no-op.
```

File: core/contracts/oracle/implementation/Voting.sol [hunk 3]
```diff
@@ -339,10 +339,25 @@ contract Voting is Testable, Ownable, OracleInterface, VotingInterface, Encrypte
         emit VoteCommitted(msg.sender, currentRoundId, identifier, time);
     }
 
+    /**
+     * @notice Snapshot the current round's token balances and lock in the inflation rate and GAT.
```
Comment (mrice32) [inline core/contracts/oracle/implementation/Voting.sol:343]: nit: we should probably note that this function will revert unless it's called during the reveal period.

Comment (chrismaree) [reply core/contracts/oracle/implementation/Voting.sol:343]: sure & added.

File: core/contracts/oracle/implementation/Voting.sol (hunk 1 above)
Comment (nicholaspai) [reply core/contracts/oracle/implementation/Voting.sol:394]: So do you anticipate in production that UMA calls `snapshotCurrentRound` before each reveal round, but we don't technically enforce this via code?

**Pull request merged**